nsq/reader.py

```diff
@@ -513,6 +519,11 @@ def _on_connection_ready(self, conn, **kwargs):
             # only send RDY 1 if we're not in backoff (some other conn
             # should be testing the waters)
             # (but always send it if we're the first)
+            if preexisting_conns and self.total_rdy == self.max_in_flight:
+                # Preexisting connections might be using all possible RDY, preventing us
+                # from sending RDY 1 on the new connection.  In that case, we force a
+                # preexisting connection to update RDY to the new, lower connection_max_in_flight.
+                self._maybe_update_rdy(preexisting_conns[0])
```

Comment:
I think you're right that it's not safe.  We need to guarantee that the conn we select has `RDY` greater than the new `connection_max_in_flight()`.  One can be shown to exist at this point, on the assumption that we're not in the abnormal regime where the connection count, counting the new conn, is greater than `max_in_flight`.  If we are in the abnormal regime, trying to set a non-zero `RDY` on the new conn is going to fail and we have to rely on periodic redistribution to give `RDY` to the new conn.

Just resetting `RDY` on all the connections, as you suggest, would simplify the logic.